File: backend-go/projects/gateway/internal/gatewaydispatch/firstbytedeadline.go
package gatewaydispatch

import (
	"context"
	"sync"
	"time"
)

// First-byte deadline decision machinery, migrated from
// upstream/first-byte-deadline.ts.

// FirstByteDeadlineAction mirrors the action union.
type FirstByteDeadlineAction string

const (
	FirstByteDeadlineActionContinue FirstByteDeadlineAction = "continue"
	FirstByteDeadlineActionAbort    FirstByteDeadlineAction = "abort"
)

// FirstByteDeadlineDecisionInput mirrors FirstByteDeadlineDecisionInput.
type FirstByteDeadlineDecisionInput struct {
	ElapsedMs int64
	TimeoutMs int64
	// Transport is 'stream' | 'non_stream'.
	Transport string
}

// FirstByteDeadlineHandler mirrors FirstByteDeadlineHandler. Node allows an
// async decision; in Go the hook is invoked synchronously inside the
// deadline goroutine and the transport aborts on its return value. The
// Node late-decision release path (decision.then(notify)) therefore has no
// pending tail here: the decision is complete before the abort applies.
type FirstByteDeadlineHandler func(input FirstByteDeadlineDecisionInput) FirstByteDeadlineAction

// NowMs is the wall-clock source for deadline decisions; tests override it.
var NowMs = func() int64 { return time.Now().UnixMilli() }

// ObservedFirstBytePendingRead mirrors observeFirstBytePendingRead: it keeps
// the settlement time of a pending read so a read that settled before the
// deadline can supersede the routing decision.
type ObservedFirstBytePendingRead[T any] struct {
	outcome chan readOutcome[T]

	mu          sync.Mutex
	settled     bool
	settledAtMs int64
}

type readOutcome[T any] struct {
	result T
	err    error
}

// ObserveFirstBytePendingRead mirrors observeFirstBytePendingRead.
func ObserveFirstBytePendingRead[T any](pendingRead func() (T, error)) *ObservedFirstBytePendingRead[T] {
	observed := &ObservedFirstBytePendingRead[T]{outcome: make(chan readOutcome[T], 1)}
	go func() {
		result, err := pendingRead()
		now := NowMs()
		observed.mu.Lock()
		observed.settled = true
		observed.settledAtMs = now
		observed.mu.Unlock()
		observed.outcome <- readOutcome[T]{result: result, err: err}
	}()
	return observed
}

// IsSettled mirrors isSettled().
func (o *ObservedFirstBytePendingRead[T]) IsSettled() bool {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.settled
}

// SettledAtMs mirrors settledAtMs().
func (o *ObservedFirstBytePendingRead[T]) SettledAtMs() (int64, bool) {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.settledAtMs, o.settled
}

// Await resolves the pending read (Node: await pendingRead.promise).
func (o *ObservedFirstBytePendingRead[T]) Await() (T, error) {
	outcome := <-o.outcome
	// Keep the outcome available for repeated Await calls.
	o.outcome = make(chan readOutcome[T], 1)
	o.outcome <- outcome
	return outcome.result, outcome.err
}

// FirstByteDeadlineDecisionResult mirrors the decision result union.
type FirstByteDeadlineDecisionResult[T any] struct {
	// Type is 'read' | 'action' | 'response_precommit_deadline'.
	Type string

	// read variant
	Result        T
	Action        FirstByteDeadlineAction
	DecisionError error
	SettledAtMs   int64

	// every variant: the read error when the read itself rejected
	Error error
}

// Decision result types.
const (
	DeadlineDecisionRead              = "read"
	DeadlineDecisionAction            = "action"
	DeadlineDecisionResponsePrecommit = "response_precommit_deadline"
)

// FirstByteDeadlineDecisionWaitOptions mirrors
// FirstByteDeadlineDecisionWaitOptions.
type FirstByteDeadlineDecisionWaitOptions struct {
	ResponsePrecommitDeadlineAtMs *int64
	OnResponsePrecommitDeadline   func()
}

// DecideFirstByteDeadlineAfterPendingRead mirrors
// decideFirstByteDeadlineAfterPendingRead: wait for the routing decision
// because it may own shared reservations or observations; the caller decides
// whether a read that settled in parallel is semantic enough to supersede
// that decision (raw activity alone is not).
func DecideFirstByteDeadlineAfterPendingRead[T any](
	pendingRead *ObservedFirstBytePendingRead[T],
	handler FirstByteDeadlineHandler,
	input FirstByteDeadlineDecisionInput,
	options FirstByteDeadlineDecisionWaitOptions,
) FirstByteDeadlineDecisionResult[T] {
	action, handlerErr := runDeadlineHandler(handler, input)
	if handlerErr != nil {
		if !pendingRead.IsSettled() {
			return FirstByteDeadlineDecisionResult[T]{Type: DeadlineDecisionAction, Error: handlerErr}
		}
		result, readErr := pendingRead.Await()
		settledAt, _ := pendingRead.SettledAtMs()
		return FirstByteDeadlineDecisionResult[T]{
			Type: DeadlineDecisionRead, Result: result, DecisionError: handlerErr,
			SettledAtMs: settledAt, Error: readErr,
		}
	}

	if options.ResponsePrecommitDeadlineAtMs == nil {
		return finishDeadlineDecision(pendingRead, action, nil)
	}

	deadlineAtMs := *options.ResponsePrecommitDeadlineAtMs
	// The Go decision settles synchronously, so the response-precommit timer
	// can only win when the wall deadline already passed (Node:
	// outcome.settledAtMs > responsePrecommitDeadlineAtMs).
	if deadlineAtMs < NowMs() {
		// Decision settled after the wall deadline: the wall wins.
		deadlineError := &GatewayResponsePrecommitDeadlineError{DeadlineAtMs: deadlineAtMs}
		notifyResponsePrecommitDeadline(options.OnResponsePrecommitDeadline)
		settledAt, settled := pendingRead.SettledAtMs()
		if settled && settledAt <= deadlineAtMs {
			result, readErr := pendingRead.Await()
			return FirstByteDeadlineDecisionResult[T]{
				Type: DeadlineDecisionRead, Result: result,
				DecisionError: deadlineError, SettledAtMs: settledAt, Error: readErr,
			}
		}
		return FirstByteDeadlineDecisionResult[T]{
			Type: DeadlineDecisionResponsePrecommit, Error: deadlineError,
		}
	}
	return finishDeadlineDecision(pendingRead, action, nil)
}

// runDeadlineHandler mirrors the try/catch around the Node handler call.
func runDeadlineHandler(handler FirstByteDeadlineHandler, input FirstByteDeadlineDecisionInput) (action FirstByteDeadlineAction, err error) {
	if handler == nil {
		return FirstByteDeadlineActionAbort, nil
	}
	defer func() {
		if recovered := recover(); recovered != nil {
			action = ""
			if panicErr, ok := recovered.(error); ok {
				err = panicErr
				return
			}
			err = &deadlineHandlerPanic{value: recovered}
		}
	}()
	return handler(input), nil
}

type deadlineHandlerPanic struct{ value any }

func (e *deadlineHandlerPanic) Error() string { return "网关首字截止决策失败" }

func finishDeadlineDecision[T any](
	pendingRead *ObservedFirstBytePendingRead[T],
	action FirstByteDeadlineAction,
	decisionErr error,
) FirstByteDeadlineDecisionResult[T] {
	if decisionErr != nil && !pendingRead.IsSettled() {
		return FirstByteDeadlineDecisionResult[T]{Type: DeadlineDecisionAction, Error: decisionErr}
	}
	if !pendingRead.IsSettled() {
		return FirstByteDeadlineDecisionResult[T]{Type: DeadlineDecisionAction, Action: action}
	}
	result, readErr := pendingRead.Await()
	settledAt, _ := pendingRead.SettledAtMs()
	return FirstByteDeadlineDecisionResult[T]{
		Type: DeadlineDecisionRead, Result: result, Action: action,
		DecisionError: decisionErr, SettledAtMs: settledAt, Error: readErr,
	}
}

func notifyResponsePrecommitDeadline(callback func()) {
	if callback == nil {
		return
	}
	defer func() { _ = recover() }()
	callback()
}

// waitForDelayMs mirrors shared/retry-policy.ts waitForRetryDelayMs with the
// abort signal translated to context cancellation.
func waitForDelayMs(ctx context.Context, delayMs int64) error {
	if delayMs <= 0 {
		return nil
	}
	timer := time.NewTimer(time.Duration(delayMs) * time.Millisecond)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func maxInt64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func minInt64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

File: migration-backup/node/final-archive/backend/src/modules/gateway/upstream/first-byte-deadline.ts
export type FirstByteDeadlineAction = 'continue' | 'abort'

export interface FirstByteDeadlineDecisionInput {
  elapsedMs: number
  timeoutMs: number
  transport: 'stream' | 'non_stream'
}

export type FirstByteDeadlineHandler = (
  input: FirstByteDeadlineDecisionInput
) => FirstByteDeadlineAction | Promise<FirstByteDeadlineAction>

export interface ObservedFirstBytePendingRead<T> {
  promise: Promise<T>
  isSettled: () => boolean
  settledAtMs: () => number | undefined
}

export type FirstByteDeadlineDecisionResult<T> =
  | {
      type: 'read'
      result: T
      action?: FirstByteDeadlineAction
      decisionError?: unknown
      settledAtMs?: number
    }
  | { type: 'action'; action: FirstByteDeadlineAction }
  | { type: 'response_precommit_deadline'; error: GatewayResponsePrecommitDeadlineError }

export interface FirstByteDeadlineDecisionWaitOptions {
  responsePrecommitDeadlineAtMs?: number
  onResponsePrecommitDeadline?: () => void
}

export class GatewayResponsePrecommitDeadlineError extends Error {
  readonly code = 'gateway_request_wall_budget_exhausted'

  constructor(readonly deadlineAtMs: number) {
    super('网关请求墙钟已到，响应尚未产生可提交的语义结果')
    this.name = 'GatewayResponsePrecommitDeadlineError'
  }
}

export function isGatewayResponsePrecommitDeadlineError(
  error: unknown
): error is GatewayResponsePrecommitDeadlineError {
  return error instanceof GatewayResponsePrecommitDeadlineError
}

export function observeFirstBytePendingRead<T>(pendingRead: Promise<T>): ObservedFirstBytePendingRead<T> {
  let settled = false
  let settledAtMs: number | undefined
  const promise = pendingRead.then(
    (result) => {
      settled = true
      settledAtMs = Date.now()
      return result
    },
    (error: unknown) => {
      settled = true
      settledAtMs = Date.now()
      throw error
    }
  )
  // The routing decision can remain pending after the read rejects. Keep the
  // observed promise handled until the caller awaits it and rethrows normally.
  void promise.catch(() => undefined)
  return {
    promise,
    isSettled: () => settled,
    settledAtMs: () => settledAtMs
  }
}

/**
 * Wait for the routing decision because it may own shared reservations or
 * observations. The caller decides whether a read that settled in parallel is
 * semantic enough to supersede that decision; raw activity alone is not.
 */
export async function decideFirstByteDeadlineAfterPendingRead<T>(
  pendingRead: ObservedFirstBytePendingRead<T>,
  handler: FirstByteDeadlineHandler | undefined,
  input: FirstByteDeadlineDecisionInput,
  options: FirstByteDeadlineDecisionWaitOptions = {}
): Promise<FirstByteDeadlineDecisionResult<T>> {
  let handlerResult: FirstByteDeadlineAction | Promise<FirstByteDeadlineAction>
  try {
    handlerResult = handler?.(input) ?? 'abort'
  } catch (error) {
    if (!pendingRead.isSettled()) {
      throw error
    }
    return {
      type: 'read',
      result: await pendingRead.promise,
      decisionError: error,
      settledAtMs: pendingRead.settledAtMs()
    }
  }

  const decision = Promise.resolve(handlerResult).then(
    (action) => ({ type: 'action' as const, action, settledAtMs: Date.now() }),
    (error: unknown) => ({ type: 'error' as const, error, settledAtMs: Date.now() })
  )
  const responsePrecommitDeadlineAtMs = options.responsePrecommitDeadlineAtMs
  let responsePrecommitTimer: NodeJS.Timeout | undefined
  const outcome = responsePrecommitDeadlineAtMs === undefined
    ? await decision
    : await Promise.race([
        decision,
        new Promise<{ type: 'response_precommit_deadline' }>((resolve) => {
          responsePrecommitTimer = setTimeout(
            () => resolve({ type: 'response_precommit_deadline' }),
            Math.max(1, responsePrecommitDeadlineAtMs - Date.now())
          )
        })
      ]).finally(() => {
        if (responsePrecommitTimer) clearTimeout(responsePrecommitTimer)
      })

  if (
    responsePrecommitDeadlineAtMs !== undefined
    && (
      outcome.type === 'response_precommit_deadline'
      || outcome.settledAtMs > responsePrecommitDeadlineAtMs
    )
  ) {
    const deadlineError = new GatewayResponsePrecommitDeadlineError(responsePrecommitDeadlineAtMs)
    notifyResponsePrecommitDeadline(options.onResponsePrecommitDeadline)
    // The local routing decision may acquire a reservation after the wall
    // deadline wins. Release again when that late decision settles.
    void decision.then(() => notifyResponsePrecommitDeadline(options.onResponsePrecommitDeadline))
    const readSettledAtMs = pendingRead.settledAtMs()
    if (
      pendingRead.isSettled()
      && readSettledAtMs !== undefined
      && readSettledAtMs <= responsePrecommitDeadlineAtMs
    ) {
      return {
        type: 'read',
        result: await pendingRead.promise,
        decisionError: deadlineError,
        settledAtMs: readSettledAtMs
      }
    }
    return { type: 'response_precommit_deadline', error: deadlineError }
  }
  if (outcome.type === 'response_precommit_deadline') {
    return {
      type: 'response_precommit_deadline',
      error: new GatewayResponsePrecommitDeadlineError(responsePrecommitDeadlineAtMs ?? Date.now())
    }
  }

  if (outcome.type === 'error') {
    if (!pendingRead.isSettled()) throw outcome.error
    return {
      type: 'read',
      result: await pendingRead.promise,
      decisionError: outcome.error,
      settledAtMs: pendingRead.settledAtMs()
    }
  }
  if (!pendingRead.isSettled()) {
    return { type: 'action', action: outcome.action }
  }
  return {
    type: 'read',
    result: await pendingRead.promise,
    action: outcome.action,
    settledAtMs: pendingRead.settledAtMs()
  }
}

function notifyResponsePrecommitDeadline(callback: (() => void) | undefined): void {
  try {
    callback?.()
  } catch {
  }
}
